soundsgood.dsp

```diff
@@ -337,38 +337,67 @@ mscomp8i(target) =
+   * Note: shelfcascade takes gains in reverse order */
+  groupGainOut  =
+    (ro.interleave(Nr_bands,N), si.bus(N))
+    : (ro.interleave(N, Nr_bands+1))
+    : par(i, N, ro.cross(Nr_bands),_);
+
+  compressor(N,prePost,strength,thresh,att,rel,knee,link) = co.peak_compression_gain_N_chan (strength,thresh,att,rel,knee,prePost,link,N);
+
+  comps(N) = (strength_array, thresh_array, att_array, rel_array, knee_array, link_array, ro.interleave (N,Nr_bands))
     : ro.interleave(Nr_bands,6+N)
-    : par(i, Nr_bands, compressor(meter(i+1),N,prePost)) ;
-  mixer = si.bus(N*Nr_bands):>si.bus(N);
+    : par(i, Nr_bands, compressor(N,prePost)) // : si.bus (N * Nr_bands)
+    : par(c, N, par(b, Nr_bands, ba.linear2db : meter(c+1, b+1)));
+
+  calcredux(N,freqs) = par(i, N, an.analyzer (6, freqs): ro.cross (Nr_bands)) : comps (N);
+
   outputGain = par(i, N, _*mscomp8i_outGain);
 
-  compressor(meter,N,prePost,strength,thresh,att,rel,knee,link) =
-    co.FFcompressor_N_chan(strength,thresh,att,rel,knee,prePost,link,meter,N);
+  meter(c,b) = _<: attach(_, (max(-40):min(0):vbargraph("v:soundsgood/t:expert/h:[6]mscomp_meter/[%c.%b][unit:dB][symbol:msredux%c.%b]mscomp redux%c %b", -3, 0)));
+
+  /* higher order low, band and hi shelf filter primitives */
+  ls3(f,g) = fi.svf.ls (f, .5, g3) : fi.svf.ls (f, .707, g3) : fi.svf.ls (f, 2, g3) with {g3 = g/3;};
+  bs3(f1,f2,g) = ls3(f1,-g) : ls3(f2,g);
+  hs3(f,g) = fi.svf.hs (f, .5, g3) : fi.svf.hs (f, .707, g3) : fi.svf.hs (f, 2, g3) with {g3 = g/3;};
 
-  //meter(i) = _<:(_, (ba.linear2db:max(-40):min(0):vbargraph("v:soundsgood/t:expert/h:[6]mscomp_meter/[%i][unit:dB]%i[tooltip: gain reduction in dB]", -3, 0))):attach;
-  meter(i) = _<:(_, (ba.linear2db:max(-40):min(0):vbargraph("v:soundsgood/t:expert/h:[6]mscomp_meter/[%i][unit:dB][tooltip: gain reduction in dB]", -3, 0))):attach;
+  /* Cascade of shelving filters to apply gain per band.
+   *
+   * `lf` : list of frequencies
+   * followed by (count(lf) +1) gain parameters
+   */
+  shelfcascade(lf) = fbus(lf), ls3(first(lf)) : sc(lf)
+  with {
+    sc((f1, f2, lf)) = fbus((f2,lf)), bs3(f1,f2) : sc((f2,lf)); // recursive pattern
+    sc((f1, f2))     = _, bs3(f1,f2) : hs3(f2);                // halting pattern
+    fbus(l)          = par(i, outputs(l), _);                  // a bus of the size of a list
+    first((x,xs))    = x;                                      // first element of a list
+  };
+
+
+  /* Cross over frequency range */
+  fl = vslider("v:soundsgood/t:expert/h:[5]mscomp/h:[1]low band/[7][symbol:mscomp_low_crossover][scale:log]low crossover", 60, 20, 4000, 1);
+  fh = vslider("v:soundsgood/t:expert/h:[5]mscomp/h:[2]high band/[7][symbol:mscomp_high_crossover][scale:log]high crossover", 8000, 5000, 20000, 1);
 
+  /* make array -- sadly cannoy use par() here. numbers need to be literal  */
+  flist = LogA(fl, fh, 7, 0), LogA(fl, fh, 7, 1), LogA(fl, fh, 7, 2), LogA(fl, fh, 7, 3), LogA(fl, fh, 7, 4), LogA(fl, fh, 7, 5), LogA(fl, fh, 7, 6)
```

Comment:
I already wrote that code:     
```faust
flist = LogArray(Nr_bands-1,fl,fh);
```

That way it also works for any number of bands.